float32 target_speed